File: projects/chess-coach/apps/harness/src/main/kotlin/harness/EngineBridge.kt
package harness

/**
 * Thin facade over [StockfishEngine] and [MaiaEngine].
 *
 * Preserves the API that [Orchestrator] depends on while delegating
 * to dedicated, timeout-aware engine wrappers backed by [UciDriver].
 */
class EngineBridge(stockfishPath: String = "stockfish") {

    private val stockfish = StockfishEngine(stockfishPath)
    private val maia = MaiaEngine()

    fun start() {
        stockfish.start()
        // Maia is started lazily on first getMove() call
    }

    fun stop() {
        stockfish.stop()
        maia.stop()
    }

    // --- Stockfish delegation ---

    data class EvalResult(val bestMove: String, val cp: Int, val isMate: Boolean, val mateIn: Int = 0)

    fun getEvaluation(fen: String, depth: Int = 15): EvalResult {
        val r = stockfish.getEvaluation(fen, depth)
        return EvalResult(r.bestMove, r.cp, r.isMate, r.mateIn)
    }

    fun checkLegality(fen: String, move: String): Boolean = stockfish.checkLegality(fen, move)

    fun getFenAfterMove(fen: String, uciMove: String): String? = stockfish.getFenAfterMove(fen, uciMove)

    // --- Maia delegation ---

    fun getMaiaMove(fen: String, difficulty: String): String = maia.getMove(fen, difficulty)
}

File: projects/chess-coach/apps/server/src/EngineBridge.ts
import { MaiaEngine } from "./MaiaEngine.ts";
import { type EvalResult, StockfishEngine } from "./StockfishEngine.ts";

export type { EvalResult };

/** The subset of EngineBridge the Orchestrator depends on (fakeable in tests). */
export interface IEngineBridge {
  getEvaluation(fen: string, depth?: number): Promise<EvalResult>;
  getFenAfterMove(fen: string, move: string): Promise<string | null>;
  getMaiaMove(fen: string, difficulty: string): Promise<string>;
}

/** Facade over Stockfish (eval/legality/FEN) + Maia (AI move). */
export class EngineBridge implements IEngineBridge {
  private readonly stockfish: StockfishEngine;
  private readonly maia: MaiaEngine;

  constructor(stockfishPath = "stockfish") {
    this.stockfish = new StockfishEngine(stockfishPath);
    this.maia = new MaiaEngine();
  }

  async start(): Promise<void> {
    await this.stockfish.start();
    // Maia starts lazily on first getMaiaMove()
  }

  stop(): void {
    this.stockfish.stop();
    this.maia.stop();
  }

  getEvaluation(fen: string, depth = 15): Promise<EvalResult> {
    return this.stockfish.getEvaluation(fen, depth);
  }

  checkLegality(fen: string, move: string): Promise<boolean> {
    return this.stockfish.checkLegality(fen, move);
  }

  getFenAfterMove(fen: string, uciMove: string): Promise<string | null> {
    return this.stockfish.getFenAfterMove(fen, uciMove);
  }

  getMaiaMove(fen: string, difficulty: string): Promise<string> {
    return this.maia.getMove(fen, difficulty);
  }
}
